### ISSUE TITLE
Faceted search indexing wrong prices and cannot be rebuilt by CRON

### ISSUE THREAD

**Issue by Hlavtox** [PrestaShop/PrestaShop#19849]
Hi, investigating my issue https://github.com/PrestaShop/PrestaShop/issues/19810 I found out several issues with the price index of Faceted search. The result is, that product sorting is not working at all. Please, this is very important. Price order is the most used one, together with stock available.

### Indexing only takes in effect specific prices
I have a product with following prices with tax:
Normal price: 1739
Specific price for group ID 4: 1394
Specific price for group ID 5: 1324

Resulting price index?
![prices](https://user-images.githubusercontent.com/6097524/85026331-d373e080-b178-11ea-8872-c982f66520e8.JPG)

### Full price regeneration is not working from link
The link for full price index rebuild is only working if clicked from module settings, not by calling the URL. I cannot do full price regeneration by CRON.

https://drive.google.com/file/d/18VkSjmvoM2kBl_g3lrsSUta-Red4vH4t/view?usp=sharing

### Additional information
* PrestaShop version: 1.7.6.2
* PHP version: 7.2

**Comment by PierreRambaud** [comment]
Hi and thanks for your report.

About specific prices, there is already a pull request about it: https://github.com/PrestaShop/ps_facetedsearch/pull/159

Can you tell us more about why the CRON is not working? What errors do you have?

And to finish, about sorting, it's failing because there is already some issues with it :sweat_smile: 

Kind regards

**Comment by Hlavtox** [comment]
@PierreRambaud 

> About specific prices, there is already a pull request about it: [PrestaShop/ps_facetedsearch#159](https://github.com/PrestaShop/ps_facetedsearch/pull/159)

Yes, that could fix the issue. For now, I commented out specific prices in my stores.

> Can you tell us more about why the CRON is not working? What errors do you have?

I cannot do full price regeneration by CRON.

https://www.domain.com/modules/ps_facetedsearch/ps_facetedsearch-price-indexer.php?token=XXXXXXXXXX works
https://www.domain.com/modules/ps_facetedsearch/ps_facetedsearch-price-indexer.php?token=XXXXXXXXXX&full=1 does not work

It works only if you click the button in the module config. It takes some time and goes through the products, but the prices are not regenerated. I believe it's due to $ajax variable set wrong.

> And to finish, about sorting, it's failing because there is already some issues with it 😅

Yes, there are, but I think there is a major design flaw. It sorts product by price_min, am I right? But it cannot done be this way, everyone will have different prices calculated and shown. 
For example, I have 2 different specific prices for my customer groups 4, 5. So, the order will be wrong, when group is set up1,2,3,4 in the context, because it sorts them by the price for group 5. And I am not even talking about specific rules for countries and whathever the hell Prestashop can be set up to 😄.

I was thinking about it _(maybe it's supposed to work that way)_ and I thought that when filtering products, min-max in the index could be used as a rough estimate of products to be selected from DB, and then you would precisely cut it by the final prices Product::priceCalculation() get you.

Not sure about price sorting though, it just has to be done by the final calculated price... Maybe a stupid question - The price for listed products is calculated anyway, so why is there even a price index?

**Comment by PierreRambaud** [comment]
> > Can you tell us more about why the CRON is not working? What errors do you have?
> 
> I cannot do full price regeneration by CRON.
> 
> https://www.domain.com/modules/ps_facetedsearch/ps_facetedsearch-price-indexer.php?token=XXXXXXXXXX works
> https://www.domain.com/modules/ps_facetedsearch/ps_facetedsearch-price-indexer.php?token=XXXXXXXXXX&full=1 does not work
> 
> It works only if you click the button in the module config. It takes some time and goes through the products, but the prices are not regenerated. I believe it's due to $ajax variable set wrong.

Can you please show me how you run your cron's script? I just try and I'm not able to reproduce the error :thinking: 

> Yes, there are, but I think there is a major design flaw. It sorts product by price_min, am I right? But it cannot done be this way, everyone will have different prices calculated and shown.
> For example, I have 2 different specific prices for my customer groups 4, 5. So, the order will be wrong, when group is set up1,2,3,4 in the context, because it sorts them by the price for group 5. And I am not even talking about specific rules for countries and whathever the hell Prestashop can be set up to smile.

> I was thinking about it (maybe it's supposed to work that way) and I thought that when filtering products, min-max in the index could be used as a rough estimate of products to be selected from DB, and then you would precisely cut it by the final prices Product::priceCalculation() get you.

> Not sure about price sorting though, it just has to be done by the final calculated price... Maybe a stupid question - The price for listed products is calculated anyway, so why is there even a price index?

I think the discussion should continue in the other issue ;)

**Comment by Hlavtox** [comment]
@PierreRambaud I want to reindex all catalog every night, to prevent errors.

1. Go to ps_layered_price_index and find one product. Set it's min and max price to zero.
2. Go to faceted search settings and copy the URL from "regenerate price index" button.
![Výstřižek](https://user-images.githubusercontent.com/6097524/85286841-a5054680-b493-11ea-8312-91fd5793ce7c.JPG)
3. Run it from your browser (or CRON, does not matter)
4. It will run for the time it should, depending on the size of the catalog.
5. Look to ps_layered_price_index, price of the product still zero.
6. Now click on the "regenerate price index" button in the module. It will work.

**Comment by PierreRambaud** [comment]
@Hlavtox I did not manage to reproduce the issue.
From my terminal:
```bash
[3.6.8] got@got:~$ curl  'http://ps-develop.localhost/modules/ps_facetedsearch/ps_facetedsearch-price-indexer.php?token=22641c7d71&full=1'
19
```

Everything is ok in the database:
![image](https://user-images.githubusercontent.com/1462701/85287200-38d71280-b494-11ea-868f-c2b8e1f97274.png)

**Comment by Hlavtox** [comment]
@PierreRambaud 
https://drive.google.com/file/d/18VkSjmvoM2kBl_g3lrsSUta-Red4vH4t/view?usp=sharing

**Comment by PierreRambaud** [comment]
> @PierreRambaud
> https://drive.google.com/file/d/18VkSjmvoM2kBl_g3lrsSUta-Red4vH4t/view?usp=sharing

Coming back from vacation, I will try to have a look as soon as possible.

**Comment by PierreRambaud** [comment]
Ohhh, regarding the code, the ajax parameter is mandatory :thinking:

**Comment by Hlavtox** [comment]
@PierreRambaud I updated the issue description for clarity. ;-)

**Comment by PierreRambaud** [comment]
@Hlavtox Tell me if everything is ok with https://github.com/PrestaShop/ps_facetedsearch/pull/177 ? :)

**Comment by Hlavtox** [comment]
@PierreRambaud No difference. It indexes only first 100 products.
Without your PR, it outputs 1 at the end, with your PR, it outputs ID of the last product indexed.

Make 500 products, index them, then go to DB, set zeroes everywhere. If you then run it by link, only first 100 are indexed. If you click button, all of them.

**Comment by PierreRambaud** [comment]
@Hlavtox I tried with more than 100 products, and all products are properly indexed :/

**Comment by Hlavtox** [comment]
@PierreRambaud Tried your PR with the updated changes. Outputs number of products to be processed.
Still indexes only first 100 products.

The second pass of indexPrices is called without $smart = true enabled and it makes the indexing not to work.

Adding `$smart = true; `after `private function indexPrices($cursor = 0, $full = false, $ajax = false, $smart = false) { `seems to make it work properly in both link and ajax mode.

**Comment by PierreRambaud** [comment]
> @PierreRambaud Tried your PR with the updated changes. Outputs number of products to be processed.
> Still indexes only first 100 products.
> 
> The second pass of indexPrices is called without $smart = true enabled and it makes the indexing not to work.
> 
> Adding `$smart = true; `after `private function indexPrices($cursor = 0, $full = false, $ajax = false, $smart = false) { `seems to make it work properly in both link and ajax mode.

Nice catch :+1: 
Could you have a new check?

**Comment by Hlavtox** [comment]
@PierreRambaud Works properly now!!! 
All zeroes, HTTP -> fine
All, zeroes, AJAX-> fine
Empty table, HTTP -> fine
Empty table, AJAX -> fine

You can merge it 🚀

**Comment by PierreRambaud** [comment]
> @PierreRambaud Works properly now!!!
> All zeroes, HTTP -> fine
> All, zeroes, AJAX-> fine
> Empty table, HTTP -> fine
> Empty table, AJAX -> fine
> 
> You can merge it

Hell yeah!

**Comment by danieledexter** [comment]
Hello everyone, I am writing here because I have encountered a similar problem.

I'm new here and i don't know if it's the right place to write.
If necessary I will open a new issue.

However my problem is that the pr_layered_price_index table has all prices = 0.
I believe this is the table used for sorting the list of products, right?

In my site I have entered all the products with price = 0 and then I have entered a specific price for each user.

I've modified my code to not show products with price = 0, this is because I wanted the final user to see only the items available to him for purchase.


I realized that the filters are not correct and if I order by increasing price at first sight the product grid is empty and to see the rest of the products I have to advance the page.

My module is at version 3.8.0 and I use Prestashop 1.7.8.6

How can I correctly value the table with specific prices per user? Is it possible? Is it also possible not to consider products with price = 0, so as not to have holes ?

Thank you all for your attention
